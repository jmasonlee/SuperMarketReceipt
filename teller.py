from model_objects import Offer
from receipt import Receipt


class Teller:

    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    def add_special_offer(self, offer_type, product, argument):
        self.offers[product] = Offer(offer_type, product, argument)

    def checks_out_articles_from(self, the_cart, receipt=None):
        #Begin Common Code
        # receipt = self.calculate_total_charges(the_cart)
        assert receipt is not None

        expected_receipt = Receipt()
        for item in the_cart.items:
            expected_receipt.add_product(item.product, item.quantity, self.catalog.unit_price(item.product), item.quantity * self.catalog.unit_price(item.product))
        assert str(receipt.items) == str(expected_receipt.items)
        ### EXIT CRITERION: receipt is not present
        ### EXIT CRITERION: receipt needs to have values for the items being purchased
        #End Common Code Note: handle_offers could also contain common code

        the_cart.handle_offers(receipt, self.offers, self.catalog)

        return receipt

    def calculate_total_charges(self, the_cart):
        receipt = Receipt()
        product_quantities = the_cart.items
        for pq in product_quantities:
            p = pq.product
            quantity = pq.quantity
            unit_price = self.catalog.unit_price(p)
            price = quantity * unit_price
            # price = 0.00 #quantity * unit_price
            receipt.add_product(p, quantity, unit_price, price)
        return receipt
