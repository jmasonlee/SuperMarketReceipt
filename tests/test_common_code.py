import pytest
from approvaltests import verify

from model_objects import Product, ProductUnit
from shopping_cart import ShoppingCart
from teller import Teller
from tests.fake_catalog import FakeCatalog


def test_common_code_empty_catalog():
    teller = Teller(FakeCatalog())
    verify(teller.common_code(ShoppingCart()))


def test_item_not_in_catalog():
    teller = Teller(FakeCatalog())

    cart = ShoppingCart()
    cart.add_item(Product("toothbrush", ProductUnit.EACH))

    with pytest.raises(KeyError) as exception:
        teller.common_code(cart)
    verify(exception)


def test_checkout_everything_in_the_catalog():
    catalog = FakeCatalog()
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apple = Product("apple", ProductUnit.KILO)
    catalog.add_product(toothbrush, 2.00)
    catalog.add_product(apple, 1.50)

    teller = Teller(catalog)

    cart = ShoppingCart()
    cart.add_item(toothbrush)
    cart.add_item(apple)

    verify(teller.common_code(cart))


def test_checkout_one_item_from_catalog():
    catalog = FakeCatalog()
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apple = Product("apple", ProductUnit.KILO)
    catalog.add_product(toothbrush, 2.00)
    catalog.add_product(apple, 1.50)

    teller = Teller(catalog)

    cart = ShoppingCart()
    cart.add_item(toothbrush)

    verify(teller.common_code(cart))
