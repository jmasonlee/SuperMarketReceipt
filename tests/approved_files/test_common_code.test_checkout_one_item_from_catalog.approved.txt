
            Receipt:
            
            Items:[toothbrush  ProductUnit.EACH 1.0 2.0 2.0]
   
            
            Discounts:[]
   
            
            Total:2.0
   
        
        
